stackslib/src/chainstate/stacks/boot/signers_voting_tests.rs

```diff
@@ -0,0 +1,435 @@
+#[test]
+fn vote_for_aggregate_public_key_in_first_block() {
+    let stacker_1 = TestStacker::from_seed(&[3, 4]);
+    let stacker_2 = TestStacker::from_seed(&[5, 6]);
+    let observer = TestEventObserver::new();
+
+    let signer = key_to_stacks_addr(&stacker_1.signer_private_key).to_account_principal();
+
+    let (mut peer, mut test_signers, latest_block_id, current_reward_cycle) = prepare_signers_test(
+        function_name!(),
+        vec![(signer, 1000)],
+        Some(vec![&stacker_1, &stacker_2]),
+        Some(&observer),
+    );
+
+    // create vote txs
+
+    let signer_nonce = 0;
+    let signer_key = &stacker_1.signer_private_key;
+    let signer_address = key_to_stacks_addr(signer_key);
+    let signer_principal = PrincipalData::from(signer_address);
+    let cycle_id = current_reward_cycle;
+
+    let signer_index = get_signer_index(&mut peer, latest_block_id, signer_address);
+
+    let aggregate_public_key: Point = Point::new();
+    let aggreagte_public_key_value =
+        Value::buff_from(aggregate_public_key.compress().data.to_vec())
+            .expect("Failed to serialize aggregate public key");
+
+    let txs = vec![
+        // cast a vote for the aggregate public key
+        make_signers_vote_for_aggregate_public_key(
+            signer_key,
+            signer_nonce,
+            signer_index,
+            &aggregate_public_key,
+            0,
+        ),
+        // cast the vote twice
+        make_signers_vote_for_aggregate_public_key(
+            signer_key,
+            signer_nonce + 1,
+            signer_index,
+            &aggregate_public_key,
+            0,
+        ),
+    ];
+
+    //
+    // vote in the first burn block of prepare phase
+    //
+    let blocks_and_sizes = nakamoto_tenure(&mut peer, &mut test_signers, vec![txs], signer_key);
+
+    // check the last two txs in the last block
+    let block = observer.get_blocks().last().unwrap().clone();
+    let receipts = block.receipts.as_slice();
+    assert_eq!(receipts.len(), 4);
+    // ignore tenure change tx
+    // ignore tenure coinbase tx
+
+    // first vote should succeed
+    let tx1 = &receipts[receipts.len() - 2];
+    assert_eq!(
+        tx1.result,
+        Value::Response(ResponseData {
+            committed: true,
+            data: Box::new(Value::Bool(true))
+        })
+    );
+
+    // second vote should fail with duplicate vote error
+    let tx2 = &receipts[receipts.len() - 1];
+    assert_eq!(
+        tx2.result,
+        Value::Response(ResponseData {
+            committed: false,
+            data: Box::new(Value::UInt(10006)) // err-duplicate-vote
+        })
+    );
+}
+
+/// In this test case, Alice votes in the first block of the last tenure of the prepare phase.
+/// Bob votes in the second block of that tenure.
+/// Alice can vote successfully.
+/// Bob is out of the voting window.
+#[test]
+fn vote_for_aggregate_public_key_in_last_block() {
```

Comment:
@kantai aren't we able to vote for DKG at any point in a reward cycle ? @friedger

Reply:
My understanding was that we need to know the number of slots. This can happen only in the prepare phase. We check that the cycle was set as and registered in .signers.